### ISSUE TITLE
Is TimeDistributed redundant?

### ISSUE THREAD

**Issue by tinkermachine** [fchollet/deep-learning-with-python-notebooks#76]
I have attempted to rework the Climate prediction example by Chollet in his book, Deep Learning with Python. Instead of simply predicting 1 variable 24 hours in the future (i.e. delay of 144 timesteps), I want to predict a sequence of 720 timesteps, 144 timesteps in the future. That is I want 5 days hours worth of predictions, 24 hours from now.

My many-to-many model is as follows:

1 hidden GRU layer of 16 units with return_sequences=True. Input to LSTM is (None, 720, 14). i.e. timesteps is 720 and dimensions is 14. return_sequences=true for GRU layer so output of this layer is (None,720,16) as expected.

Output layer is simply a Dense layer of 1 neuron which outputs a shape (None, 720, 1)

I have notices regardless of whether I use regular Dense (1) layer or a TimeDistributed(Dense(1)) layer as the output, the loss and predictions are exactly the same. Also, the output shape from the Dense layer is (None, 720, 1) regardless of whether I use TimeDistributed(Dense(1)) or regular Dense(1).

My questions is if this is expected and if so, in which scenario should I expect to use the TimeDistributed functionality?

My gist can be found here: https://gist.github.com/tinkermachine/e87b6730a71cecdbebc5dfb25d810c9c

**Comment by rickyzhang82** [comment]
To dense layer (a.k.a fully connected layer). `TimeDistributed ` really doesn't seem to do what it claims in the Keras document that "apply the layer **independently** on the input".

See my test case below. I'd expect that the first one `time_dist_lstm ` would have 1000 * 33 =33, 000 trainable parameters.

```Python
from keras.layers import LSTM
from keras.layers import Dense
from keras.layers import TimeDistributed
from keras.models import Sequential

def test_lstm_td_model(lookback, num_features, should_time_distributed=False):
    model = Sequential()
    model.add(LSTM(32, input_shape=(lookback, num_features), 
                   return_sequences=True, stateful=False))
    if should_time_distributed:
        model.add(TimeDistributed(Dense(1)))
    else:
        model.add(Dense(1))

    model.compile(optimizer='rmsprop',
                  loss='binary_crossentropy',
                  metrics=['acc'])
    return model  

time_dist_lstm = test_lstm_td_model(1000, 5, should_time_distributed=True)
time_dist_lstm.summary()
no_time_dist_lstm = test_lstm_td_model(1000, 5, should_time_distributed=False)
no_time_dist_lstm.summary()
```
The output
```
Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
lstm_8 (LSTM)                (None, 1000, 32)          4864      
_________________________________________________________________
dense_8 (Dense)              (None, 1000, 1)           33        
=================================================================
Total params: 4,897
Trainable params: 4,897
Non-trainable params: 0
_________________________________________________________________
Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
lstm_9 (LSTM)                (None, 1000, 32)          4864      
_________________________________________________________________
time_distributed_9 (TimeDist (None, 1000, 1)           33        
=================================================================
Total params: 4,897
Trainable params: 4,897
Non-trainable params: 0
```

In any case, if you want to predict two forecast - 1 day ahead and 5 days ahead. You really don't need to  use `TimeDistributed ` on `Dense` layer. You can output Dense(2) and set return_sequence= False. And fit 1 day ahead and 5 days ahead together.

There is an alternative way I have seen in practice that can drastically improve training time and prediction accuracy for multi-time step forecast:
1. Two separate model for 1 day ahead and 5 days ahead model. But the model shared the same structure.
1. First, train the model that Forecast 1 day ahead.
1. Use 1 day ahead weight to initialize 5 day ahead model. Use 1 day ahead forecast value as part of the input for 5 day ahead model.

For sure, you need to initialize something in 1 day ahead model at the beginning.
